pkg/services/ngalert/schedule/sequence.go

```diff
@@ -0,0 +1,121 @@
+package schedule
+
+import (
+	"cmp"
+	"slices"
+	"strings"
+
+	models "github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// sequence represents a chain of rules that should be evaluated in order.
+// It is a convience type that wraps readyToRunItem as an indicator of what
+// is being represented.
+type sequence readyToRunItem
+
+type groupKey struct {
+	folderTitle string
+	folderUID   string
+	groupName   string
+}
+
+// buildSequences organizes rules into evaluation sequences where rules in the same group
+// are chained together. The first rule in each group will trigger the evaluation of subsequent
+// rules in that group through the afterEval callback.
+//
+// For example, if we have rules A, B, C in group G1 and rules D, E in group G2:
+// - A will have afterEval set to evaluate B
+// - B will have afterEval set to evaluate C
+// - D will have afterEval set to evaluate E
+//
+// The function returns a slice of sequences, where each sequence represents a chain of rules
+// that should be evaluated in order.
+//
+// NOTE: This currently only chains rules in imported groups.
+func (sch *schedule) buildSequences(items []readyToRunItem, runJobFn func(next readyToRunItem, prev ...readyToRunItem) func()) []sequence {
+	// Step 1: Group rules by their folder and group name
+	groups := map[groupKey][]readyToRunItem{}
+	var keys []groupKey
+	for _, item := range items {
+		g := groupKey{
+			folderTitle: item.folderTitle,
+			folderUID:   item.rule.NamespaceUID,
+			groupName:   item.rule.RuleGroup,
+		}
+		i, ok := groups[g]
+		if !ok {
+			keys = append(keys, g)
+		}
+		groups[g] = append(i, item)
+	}
+
+	// Step 2: Sort group keys to ensure consistent ordering
+	slices.SortFunc(keys, func(a, b groupKey) int {
+		return cmp.Or(
+			cmp.Compare(a.folderTitle, b.folderTitle),
+			cmp.Compare(a.folderUID, b.folderUID),
+			cmp.Compare(a.groupName, b.groupName),
+		)
+	})
+
+	// Step 3: Build evaluation sequences for each group
+	result := make([]sequence, 0, len(items))
+	for _, key := range keys {
+		groupItems := groups[key]
+
+		if sch.shouldEvaluateSequentially(groupItems) {
+			result = append(result, sch.buildSequence(key, groupItems, runJobFn))
+			continue
+		}
+
+		for _, item := range groupItems {
+			result = append(result, sequence(item))
+		}
+	}
+
+	// sort the sequences by UID
+	slices.SortFunc(result, func(a, b sequence) int {
+		return strings.Compare(a.rule.UID, b.rule.UID)
+	})
+
+	return result
+}
+
+func (sch *schedule) buildSequence(groupKey groupKey, groupItems []readyToRunItem, runJobFn func(next readyToRunItem, prev ...readyToRunItem) func()) sequence {
+	slices.SortFunc(groupItems, func(a, b readyToRunItem) int {
```

Comment:
NIT: Could be internally defensive here and check for `len(groupItems) < 2`. Not a hard requirement as technically we ensure this in the caller.

Reply:
I originally left it out since this is a private method, but it can't hurt to add.